File: hard/solveNQueens.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> solutions = new ArrayList<>();
        int[] queens = new int[n];
        boolean[] columns = new boolean[n];
        boolean[] diagonals1 = new boolean[2 * n - 1];
        boolean[] diagonals2 = new boolean[2 * n - 1];
        backtrack(solutions, queens, n, 0, columns, diagonals1, diagonals2);
        return solutions;
    }

    private void backtrack(List<List<String>> solutions, int[] queens, int n, int row,
                           boolean[] columns, boolean[] diagonals1, boolean[] diagonals2) {
        if (row == n) {
            solutions.add(generateBoard(queens, n));
        } else {
            for (int i = 0; i < n; i++) {
                int diag1 = row - i + n - 1;
                int diag2 = row + i;
                if (columns[i] || diagonals1[diag1] || diagonals2[diag2]) {
                    continue;
                }
                queens[row] = i;
                columns[i] = true;
                diagonals1[diag1] = true;
                diagonals2[diag2] = true;
                backtrack(solutions, queens, n, row + 1, columns, diagonals1, diagonals2);
                queens[row] = -1;
                columns[i] = false;
                diagonals1[diag1] = false;
                diagonals2[diag2] = false;
            }
        }
    }

    private List<String> generateBoard(int[] queens, int n) {
        List<String> board = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            char[] row = new char[n];
            for (int k = 0; k < n; k++) {
                row[k] = '.';
            }
            row[queens[i]] = 'Q';
            board.add(new String(row));
        }
        return board;
    }
}

File: hard/solveNQueens.py
class Solution:
    def solveNQueens(self, n):
        solutions = []
        queens = [-1] * n
        columns = [False] * n
        diagonals1 = [False] * (2 * n - 1)
        diagonals2 = [False] * (2 * n - 1)
        self.backtrack(solutions, queens, n, 0, columns, diagonals1, diagonals2)
        return solutions

    def backtrack(self, solutions, queens, n, row, columns, diagonals1, diagonals2):
        if row == n:
            solutions.append(self.generateBoard(queens, n))
        else:
            for i in range(n):
                diag1 = row - i + n - 1
                diag2 = row + i
                if columns[i] or diagonals1[diag1] or diagonals2[diag2]:
                    continue
                queens[row] = i
                columns[i] = True
                diagonals1[diag1] = True
                diagonals2[diag2] = True
                self.backtrack(solutions, queens, n, row + 1, columns, diagonals1, diagonals2)
                queens[row] = -1
                columns[i] = False
                diagonals1[diag1] = False
                diagonals2[diag2] = False

    def generateBoard(self, queens, n):
        board = []
        for i in range(n):
            row = ['.'] * n
            row[queens[i]] = 'Q'
            board.append(''.join(row))
        return board
